File: ecentric_workspace/platform/esign/ui/coords.js
/* Copyright (c) 2026, eCentric and contributors
 * Canonical placement coordinate math for the bundled PDF placement editor.
 * This is a 1:1 mirror of esign/coords.py. Canonical = TOP-LEFT origin, PDF points,
 * unrotated mediabox. Viewport = PDF.js rendered (rotated) canvas, CSS px at scale S.
 * The backend re-validates authoritatively; this is a deterministic display transform.
 * Exposed as window.ECoords (browser) and module.exports (node test).
 */
(function (root) {
  "use strict";
  var ROT = [0, 90, 180, 270];

  function normalizeRotation(r) {
    var v = parseInt(r || 0, 10);
    if (isNaN(v)) return 0;
    v = ((v % 360) + 360) % 360;
    return ROT.indexOf(v) >= 0 ? v : 0;
  }
  function renderedPageSize(pagePt, rotation) {
    var w = +pagePt[0], h = +pagePt[1], r = normalizeRotation(rotation);
    return (r === 90 || r === 270) ? [h, w] : [w, h];
  }
  function canonicalPointToRendered(cx, cy, pagePt, rotation) {
    var w = +pagePt[0], h = +pagePt[1], r = normalizeRotation(rotation);
    if (r === 0) return [cx, cy];
    if (r === 90) return [h - cy, cx];
    if (r === 180) return [w - cx, h - cy];
    return [cy, w - cx]; // 270
  }
  function renderedPointToCanonical(rx, ry, pagePt, rotation) {
    var w = +pagePt[0], h = +pagePt[1], r = normalizeRotation(rotation);
    if (r === 0) return [rx, ry];
    if (r === 90) return [ry, h - rx];
    if (r === 180) return [w - rx, h - ry];
    return [w - ry, rx]; // 270
  }
  function rectFromCorners(ax, ay, bx, by) {
    return { x: Math.min(ax, bx), y: Math.min(ay, by),
             width: Math.abs(bx - ax), height: Math.abs(by - ay) };
  }
  function round3(v) { return Math.round(v * 1000) / 1000; }

  function viewportRectToCanonical(rectPx, pagePt, scale, rotation) {
    var s = +scale;
    if (!(s > 0)) throw new Error("scale must be > 0");
    var rx1 = rectPx.x / s, ry1 = rectPx.y / s;
    var rx2 = (rectPx.x + rectPx.width) / s, ry2 = (rectPx.y + rectPx.height) / s;
    var c1 = renderedPointToCanonical(rx1, ry1, pagePt, rotation);
    var c2 = renderedPointToCanonical(rx2, ry2, pagePt, rotation);
    var o = rectFromCorners(c1[0], c1[1], c2[0], c2[1]);
    return { x: round3(o.x), y: round3(o.y), width: round3(o.width), height: round3(o.height) };
  }
  function canonicalRectToViewport(rectPt, pagePt, scale, rotation) {
    var s = +scale;
    if (!(s > 0)) throw new Error("scale must be > 0");
    var cx1 = rectPt.x, cy1 = rectPt.y, cx2 = rectPt.x + rectPt.width, cy2 = rectPt.y + rectPt.height;
    var r1 = canonicalPointToRendered(cx1, cy1, pagePt, rotation);
    var r2 = canonicalPointToRendered(cx2, cy2, pagePt, rotation);
    var o = rectFromCorners(r1[0] * s, r1[1] * s, r2[0] * s, r2[1] * s);
    return { x: round3(o.x), y: round3(o.y), width: round3(o.width), height: round3(o.height) };
  }
  function isWithinPage(rectPt, pagePt, tol) {
    tol = tol === undefined ? 1.0 : tol;
    var w = +pagePt[0], h = +pagePt[1];
    if (!(rectPt.width > 0) || !(rectPt.height > 0)) return false;
    return rectPt.x >= -tol && rectPt.y >= -tol &&
           (rectPt.x + rectPt.width) <= (w + tol) && (rectPt.y + rectPt.height) <= (h + tol);
  }

  var API = { normalizeRotation: normalizeRotation, renderedPageSize: renderedPageSize,
    canonicalPointToRendered: canonicalPointToRendered,
    renderedPointToCanonical: renderedPointToCanonical,
    viewportRectToCanonical: viewportRectToCanonical,
    canonicalRectToViewport: canonicalRectToViewport, isWithinPage: isWithinPage };
  if (typeof module !== "undefined" && module.exports) module.exports = API;
  root.ECoords = API;
})(typeof window !== "undefined" ? window : this);

File: ecentric_workspace/platform/esign/coords.py
# Copyright (c) 2026, eCentric and contributors
"""Canonical placement coordinate system for the bundled PDF placement editor.

This module is deliberately frappe-free and side-effect-free so the conversion is
deterministic and fully unit-testable, and so the SAME math can be mirrored 1:1 in the
browser editor (esign/ui/pdf_placement_editor.html).

CANONICAL SYSTEM (the only thing persisted / validated by the backend):
  * origin = TOP-LEFT of the UNROTATED PDF page (matches package.validate_placement_geometry
    which validates x/y/width/height in points against the mediabox with a top-left origin);
  * +x = right, +y = down; units = PDF points (1/72 inch);
  * page dimensions = the page's UNROTATED mediabox (width_pt, height_pt);
  * page_index is 1-based.

VIEWPORT SYSTEM (what PDF.js draws on screen):
  * PDF.js renders the page already rotated by the page /Rotate value (0/90/180/270,
    clockwise) at a scale factor S (CSS px per rendered point);
  * viewport origin = TOP-LEFT of the rendered canvas, +x right, +y down, units = CSS px.

The editor draws rectangles in viewport px; we convert to canonical points before sending
to the backend, and the backend re-validates authoritatively. Frontend numbers are never
trusted for authorization or bounds - this is purely a deterministic display transform.
"""

_ROT = (0, 90, 180, 270)


def normalize_rotation(rotation):
    """Coerce any PDF /Rotate into {0,90,180,270}. Non-multiples of 90 fail closed to 0."""
    try:
        r = int(rotation or 0) % 360
    except (TypeError, ValueError):
        return 0
    return r if r in _ROT else 0


def rendered_page_size(page_pt, rotation):
    """(rendered_width_pt, rendered_height_pt) for the on-screen rotated page."""
    w, h = float(page_pt[0]), float(page_pt[1])
    return (h, w) if normalize_rotation(rotation) in (90, 270) else (w, h)


def canonical_point_to_rendered(cx, cy, page_pt, rotation):
    """Map a canonical (unrotated, top-left, points) point to rendered (rotated, top-left,
    points) coordinates. Clockwise display rotation, matching PDF /Rotate semantics."""
    w, h = float(page_pt[0]), float(page_pt[1])
    r = normalize_rotation(rotation)
    if r == 0:
        return (cx, cy)
    if r == 90:
        return (h - cy, cx)
    if r == 180:
        return (w - cx, h - cy)
    return (cy, w - cx)  # 270


def rendered_point_to_canonical(rx, ry, page_pt, rotation):
    """Inverse of canonical_point_to_rendered."""
    w, h = float(page_pt[0]), float(page_pt[1])
    r = normalize_rotation(rotation)
    if r == 0:
        return (rx, ry)
    if r == 90:
        return (ry, h - rx)
    if r == 180:
        return (w - rx, h - ry)
    return (w - ry, rx)  # 270


def _rect_from_corners(ax, ay, bx, by):
    x = min(ax, bx)
    y = min(ay, by)
    return {"x": x, "y": y, "width": abs(bx - ax), "height": abs(by - ay)}


def viewport_rect_to_canonical(rect_px, page_pt, scale, rotation):
    """Convert a viewport-pixel rectangle {x,y,width,height} (top-left) into a canonical
    point rectangle. `scale` = CSS px per rendered point (>0). Rounds to 3 decimals so the
    same input always yields the same persisted value (determinism for idempotency/hashing).
    """
    s = float(scale)
    if s <= 0:
        raise ValueError("scale must be > 0")
    # viewport px -> rendered points
    rx1 = float(rect_px["x"]) / s
    ry1 = float(rect_px["y"]) / s
    rx2 = (float(rect_px["x"]) + float(rect_px["width"])) / s
    ry2 = (float(rect_px["y"]) + float(rect_px["height"])) / s
    # rendered points -> canonical points (transform both corners, then bound)
    c1 = rendered_point_to_canonical(rx1, ry1, page_pt, rotation)
    c2 = rendered_point_to_canonical(rx2, ry2, page_pt, rotation)
    out = _rect_from_corners(c1[0], c1[1], c2[0], c2[1])
    return {k: round(v, 3) for k, v in out.items()}


def canonical_rect_to_viewport(rect_pt, page_pt, scale, rotation):
    """Inverse of viewport_rect_to_canonical - used to render existing persisted placements
    back onto the rotated viewport."""
    s = float(scale)
    if s <= 0:
        raise ValueError("scale must be > 0")
    cx1 = float(rect_pt["x"])
    cy1 = float(rect_pt["y"])
    cx2 = cx1 + float(rect_pt["width"])
    cy2 = cy1 + float(rect_pt["height"])
    r1 = canonical_point_to_rendered(cx1, cy1, page_pt, rotation)
    r2 = canonical_point_to_rendered(cx2, cy2, page_pt, rotation)
    out = _rect_from_corners(r1[0] * s, r1[1] * s, r2[0] * s, r2[1] * s)
    return {k: round(v, 3) for k, v in out.items()}


def is_within_page(rect_pt, page_pt, tol=1.0):
    """True if a canonical rectangle stays inside the unrotated mediabox (matches the
    backend rule). Zero/negative-area rectangles are rejected."""
    w, h = float(page_pt[0]), float(page_pt[1])
    x, y = float(rect_pt["x"]), float(rect_pt["y"])
    wd, ht = float(rect_pt["width"]), float(rect_pt["height"])
    if wd <= 0 or ht <= 0:
        return False
    return (x >= -tol and y >= -tol and (x + wd) <= (w + tol) and (y + ht) <= (h + tol))
